text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+  `std::os` provided a small "speed bump" for venturing out of guaranteed
+  mainstream platform portability.
+
+- Finally, for working in low-level and embedded contexts, we stabilized
+  `libcore`, a subset of `libstd` that excludes all OS services and allocation,
+  but *still* makes some hardware assumptions (e.g. about atomics and floating
+  point support).
+
+## Problems with the status quo
+
+The above strategy has served us fairly well in the first year since Rust 1.0,
+but it's increasingly holding us back from enhancements we'd like to make, and
+even for the needs it covers, it's suboptimal in a few ways.
+
+**Problems with `std::os`**:
+
+* The `std::os` module has submodules that correspond to a hierarchy of OS
+  types. For example, there is a `unix` submodule that applies to several
+  operating systems, but there's also a `linux` submodule with Linux-specific
+  extensions. There are a couple of problems with such an organization. Most
+  importantly, it's not at all clear how to use the module hierarchy to organize
+  features like [fixed-size atomic types][more-atomics], where the types
+  available vary in a fine-grained way based on the CPU family; [SIMD] is even
+  worse. But even just for operating systems, organizing into a hierarchy
+  becomes difficult as we gain more and more APIs, some of which are only
+  available on particular *versions* of a given operating system.
+
+* The "speed bump" for using `std::os` is minimal and easy to miss; it's just an
+  import that looks the same as any other. Moreover, it doesn't provide any help
+  with the ecosystem beyond `std`. There's no simple way to tell whether a crate
+  you're relying on is portable to the same degree as `std` is, and the `os`
+  submodule pattern has not really caught on in the wider ecosystem.
+
+* Platform-specific APIs don't live in their "natural location". The majority of
+  `std::os` works through extension traits to enhance the functionality of
+  standard primitives. For example `std::os::unix::io::AsRawFd` is a trait with
+  the `as_raw_fd` method (to extract a file descriptor). If you were to ignore
+  Windows, however, one might expect this API instead to live as a method
+  directly on types like `File`, `TcpStream`, etc. Forcing code to live in
+  `std::os` thus comes at a mild cost for both ergonomics and discoverability.
+  This problem is even worse for features like adding more atomic types or SIMD.
+
+**Problems with `libcore`/the facade**:
+
+* Embedded libraries typically wish to never use functions in the standard
+  library that abort on allocation failure (e.g. `Vec::push`). We'd like to
+  provide some way for these libraries to use and interoperate with the standard
+  collection types, but only have access to an alternative API surface (e.g. a
+  `try_push` method provided via an extension trait). It's not clear how to do
+  that with the current [facade] setup.
+
+* Kernels and embedded environments often want to
```

Comment:
[I thought this was already solved?](https://github.com/rust-lang/rfcs/issues/1364#issuecomment-236374104) Using the right codegen options removed the LLVM assertions that x86 devs were seeing by lowering float operations to intrinsics (software implementations) *without* having to modify the `core` crate. cc @steveklabnik @phil-opp Does the #rust-osdev community still recommend patching `core` to remove floats?

On the ARM Cortex-M side of embedded development, if your hardware doesn't have floating point instructions then you use, for example, the `thumbv7em-none-eabi` target instead of the "hard float" variant: `thumbv7em-none-eabihf`, which does have floating point instructions. No need to strip anything from the `core` crate.